tools/linter/adapters/missing_type_linter.py

```diff
@@ -0,0 +1,214 @@
+        assert isinstance(line, int)
+        return line
+
+    @cached_property
+    def lint_result(self) -> LintResult:
+        category = "parameter" if self.param_name else "return"
+        name = f"Missing {category} type for {self.grandfather}"
+        return LintResult(
+            char=self.column, length=self.length, line=self.line, name=name
+        )
+
+    @cached_property
+    def python_parts(self) -> tuple[str, ...]:
+        return self.python_file.python_parts
+
+    @cached_property
+    def suffix(self) -> str:
+        return f"({self.param_name}=)" if self.param_name else ""
+
+
+class MissingTypeLinter(FileLinter):
+    linter_name = "missing_type_linter"
+    description = DESCRIPTION
+    epilog = EPILOG
+    report_column_numbers = True
+
+    def __init__(self, argv: Sequence[str] | None = None) -> None:
+        super().__init__(argv)
+        add = self.parser.add_argument
+
+        help = "The paths to check"
+        add("--path", "-p", nargs="*", help=help)
+
+        help = f"Set the grandfather list (default={GRANDFATHER})"
+        add("--grandfather", "-g", default=GRANDFATHER, type=Path, help=help)
+
+        help = f"The command line for type checking (default='{TYPE_CHECK_COMMAND}')"
+        add("--type-check", "-t", default=TYPE_CHECK_COMMAND, help=help)
+
+        help = "A JSON file with the type checking results (so --type-check is ignored)"
+        add("--type-result", "-r", default=None, type=Path, help=help)
+
+        help = "Rewrite the grandfather list"
+        add("--write-grandfather", "-w", action="store_true", help=help)
+
+    def lint_all(self) -> bool:
+        if self.write_grandfather:
+            with self.args.grandfather.open("w") as fp:
+                fp.writelines(g + "\n" for g in self.grandfather)
+            return True
+        else:
+            for k in self.missing_annotations_added:
+                self._lint_file(Path(k))
+            return not self.missing_annotations_added
+
+    def _lint(self, pf: PythonFile) -> Iterator[LintResult]:
+        yield from (m.lint_result for m in self.missing_annotations_added[str(pf.path)])
+
+    @cached_property
+    def missing_annotations_added(self) -> dict[str, list[MissingAnnotation]]:
+        with self.args.grandfather.open() as fp:
+            grandfather = {i.strip() for i in fp}
+
+        def grandfathered(lm: list[MissingAnnotation]) -> list[MissingAnnotation]:
+            return [m for m in lm if m.grandfather not in grandfather]
+
+        items = self.missing_annotations.items()
+        return {k: g for k, v in items if (g := grandfathered(v))}
+
+    @cached_property
+    def grandfather(self) -> list[str]:
+        annotations = self.missing_annotations.values()
+        # TODO: investigate dupes
+        return sorted({m.grandfather for v in annotations for m in v})
+
+    @cached_property
+    def missing_annotations(self) -> dict[str, list[MissingAnnotation]]:
```

Comment:
Here's where we compute the public symbols (functions or parameters) that are missing annotations - from the `pyrefly` report, which lists every single symbol and all its annotations.